src/platform/web/lib/transport/push.ts

```diff
@@ -0,0 +1,84 @@
+import { ActivationStateMachine, LocalDevice } from 'plugins/push/pushactivation';
+
+function toBase64Url(arrayBuffer: ArrayBuffer) {
+  const buffer = new Uint8Array(arrayBuffer.slice(0, arrayBuffer.byteLength));
+  return btoa(String.fromCharCode.apply(null, Array.from(buffer)));
+}
+
+function urlBase64ToUint8Array(base64String: string) {
+  const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
+  const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');
+  const rawData = window.atob(base64);
+  const rawDataChars = [];
+  for (let i = 0; i < rawData.length; i++) {
+    rawDataChars.push(rawData[i].charCodeAt(0));
+  }
+  return Uint8Array.from(rawDataChars);
+}
+
+export async function getW3CPushDeviceDetails(machine: ActivationStateMachine) {
+  const GettingPushDeviceDetailsFailed = machine.GettingPushDeviceDetailsFailed;
+  const GotPushDeviceDetails = machine.GotPushDeviceDetails;
+  const { ErrorInfo, Defaults } = machine.rest;
+
+  const permission = await Notification.requestPermission();
+
+  if (permission !== 'granted') {
+    machine.handleEvent(
+      new GettingPushDeviceDetailsFailed(new ErrorInfo(`user denied permission to send notifications.`, 400, 40000)),
+    );
+    return;
+  }
+
+  const swUrl = machine.rest.options.pushServiceWorkerUrl;
+  if (!swUrl) {
+    machine.handleEvent(
+      new GettingPushDeviceDetailsFailed(new ErrorInfo('missing ClientOptions.pushServiceWorkerUrl', 400, 40000)),
+    );
+    return;
+  }
+
+  try {
+    const worker = await navigator.serviceWorker.register(swUrl);
+
+    machine._pushManager = worker.pushManager;
+
+    const headers = Defaults.defaultGetHeaders(machine.rest.options, { format: 'text' });
+    const appServerKey = (
+      await machine.rest.Resource!.get(machine.rest, '/push/publicVapidKey', headers, {}, null, true)
+    ).body as string;
+
+    if (!worker.active) {
+      await navigator.serviceWorker.ready;
+    }
+
+    const subscription = await worker.pushManager.subscribe({
+      userVisibleOnly: true,
+      applicationServerKey: urlBase64ToUint8Array(appServerKey),
+    });
+
+    const endpoint = subscription.endpoint;
+
+    const [p256dh, auth] = [subscription.getKey('p256dh'), subscription.getKey('auth')];
+
+    if (!p256dh || !auth) {
+      throw new ErrorInfo('Public key not found', 50000, 500);
+    }
+
+    const key = [p256dh, auth].map(toBase64Url).join(':');
+
+    const device = machine.rest.device as LocalDevice;
+    device.push.recipient = {
+      transportType: 'web',
+      targetUrl: btoa(endpoint),
+      encryptionKey: key,
+    };
+    device.persist();
+
+    machine.handleEvent(new GotPushDeviceDetails());
+  } catch (err) {
+    machine.handleEvent(
+      new GettingPushDeviceDetailsFailed(new ErrorInfo('failed to register service worker', 50000, 500, err as Error)),
```

Comment:
```suggestion
      new GettingPushDeviceDetailsFailed(new ErrorInfo('Failed to register service worker', 50000, 500, err as Error)),
```
nit: more consistent with the rest of codebase